### PR TITLE
Implement whitespace control for variables

### PR DESCRIPTION
**Pull request by funkjedi**
- [x] I've run the tests with `vendor/bin/phpunit`
- [x] None of the tests were found failing
- [x] I've seen the coverage report at `build/coverage/index.html`
- [x] Not a single line left uncovered by tests
- [x] Any coding standards issues were fixed with `vendor/bin/php-cs-fixer fix`

Builds on whitespace control introduced in #110 and discussed in #109 using implementation based on the ruby version of Liquid. Adds support for whitespace control of variables.

### CODE REVIEW

File: src/Liquid/Liquid.php [hunk 1]
```diff
@@ -52,11 +52,14 @@ class Liquid
...
 		// Tag start.
-		'TAG_START' => '(?:{%|\s*{%-)',
+		'TAG_START' => '{%',
 
 		// Tag end.
-		'TAG_END' => '(?:%}|-%}\s*)',
+		'TAG_END' => '%}',
 
 		// Variable start.
 		'VARIABLE_START' => '{{',
```
Comment (sanmai) [inline src/Liquid/Liquid.php]: `VARIABLE_START` could have used the very same approach used for `TAG_START`. Why do you think this approach is inferior?

Comment (funkjedi) [reply src/Liquid/Liquid.php]: It's not so much about inferiority as it is compatibility/simplicity. Having the start/end tags include the whitespace control:

1. Increases the complexity of customizing the start/end tags. Especially for those not familiar with regular expressions.
2. Means anyone who's already customized their start/end tags won't be able to use whitespace control unless they update them to include the new whitespace regex.

File: src/Liquid/AbstractBlock.php [hunk 2]
```diff
@@ -83,15 +89,38 @@ public function parse(array &$tokens)
...
+	protected function whitespaceHandler(&$token)
+	{
+		if (mb_strlen($token) > 2 && $token[2] === Liquid::get('WHITESPACE_CONTROL')) {
+			$previousToken = end($this->nodelist);
+			if (is_string($previousToken)) {
+				$this->nodelist[key($this->nodelist)] = rtrim($previousToken);
+			}
+		}
+
+		self::$trimWhitespace = (mb_strlen($token) > 2 && substr($token, -3, 1) === Liquid::get('WHITESPACE_CONTROL'));
```
Comment (sanmai) [inline src/Liquid/AbstractBlock.php]: I'm having hard time understanding what this function is supposed to do exactly and why. Could you elaborate in a comment?

For example, `mb_strlen` is used here and above, but on the other hand array access operator used which treats all strings as single-byte arrays. It's either `mb_strlen` or `$string[n]` and the common `strlen`, else it makes no sense to me. [A quick demo.](https://3v4l.org/8AVp9)

File: src/Liquid/Liquid.php (hunk 1 above)
Comment (sanmai) [reply src/Liquid/Liquid.php]: Makes sense, thank you.

File: src/Liquid/AbstractBlock.php [hunk 3]
```diff
@@ -111,14 +111,14 @@ public function parse(array &$tokens)
...
-		if (mb_strlen($token) > 2 && $token[2] === Liquid::get('WHITESPACE_CONTROL')) {
+		if (mb_strlen($token) > 2 && mb_substr($token, 2, 1) === Liquid::get('WHITESPACE_CONTROL')) {
 			$previousToken = end($this->nodelist);
 			if (is_string($previousToken)) {
 				$this->nodelist[key($this->nodelist)] = rtrim($previousToken);
 			}
 		}
 
-		self::$trimWhitespace = (mb_strlen($token) > 2 && substr($token, -3, 1) === Liquid::get('WHITESPACE_CONTROL'));
+		self::$trimWhitespace = (mb_strlen($token) > 2 && mb_substr($token, -3, 1) === Liquid::get('WHITESPACE_CONTROL'));
```
Comment (sanmai) [inline src/Liquid/AbstractBlock.php]: What do you think if we just use `mb_substr` here and remove `mb_strlen` altogether?

**Comment by funkjedi** [comment]
In Ruby implementation there's `parse_context` object that's passed through the callstack when parsing that maintains the state of a `trim_whitespace` boolean.
```
module Liquid
  class ParseContext
    attr_accessor :locale, :line_number, :trim_whitespace, :depth
    attr_reader :partial, :warnings, :error_mode
```
Since the PHP implementation doesn't do that the static var is just a quick and dirty way of maintaining that state with minimal reengineering of the PHP implementation.

**Comment by sanmai** [comment]
> The following {%- if 1 -%}display{% endif %} should output display.
This must be a test.

**Pull request merged**